### PR TITLE
added support for 'searchsorted'

### PR DESCRIPTION
**Pull request by mfoerste4**
CPU/OMP & GPU support for searchsorted

The (local) GPU implementation is kept simple for now, using a single thread for every value to-be-inserted performing binary search. AFAIU this is how cupy does it. It might make sense to do something more sophisticated here.

### CODE REVIEW

File: tests/integration/test_searchsorted.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+import numpy as np
+import pytest
+
+import cunumeric as num
+
+
+def compare_assert(a_np, a_num):
+    if not num.allclose(a_np, a_num):
+        print("numpy, shape " + str(a_np.shape) + ":")
```
Comment (magnatelee) [inline tests/integration/test_searchsorted.py:24]: use f-string (here and all the other places where you manually concatenate strings)

File: tests/integration/test_searchsorted.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    print("Right/Array/Sorter")
+    compare_assert(
+        a.searchsorted(v, side="right", sorter=a_argsorted),
+        a_num.searchsorted(v_num, side="right", sorter=a_num_argsorted),
+    )
+
+
+def check_dtypes():
+    np.random.seed(42)
+    check_api(generate_random((156,), np.uint8))
```
Comment (magnatelee) [inline tests/integration/test_searchsorted.py:124]: consider using pytest decorators instead of enumerating test calls like this. e.g., https://github.com/nv-legate/cunumeric/blob/branch-22.07/tests/integration/test_matmul.py#L28-L29

File: tests/integration/test_searchsorted.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+            np.random.random(size=volume) + np.random.random(size=volume) * 1j,
+            dtype=datatype,
+        )
+    else:
+        print("UNKNOWN type " + str(datatype))
+        assert False
+    return a_np.reshape(shape)
+
+
+def check_api(a=None):
```
Comment (magnatelee) [inline tests/integration/test_searchsorted.py:58]: you never call this function with `a` being `None`, so why do you make this an optional parameter?

File: tests/integration/test_searchsorted.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+def compare_assert(a_np, a_num):
+    if not num.allclose(a_np, a_num):
+        print("numpy, shape " + str(a_np.shape) + ":")
+        print(a_np)
+        print("cuNumeric, shape " + str(a_num.shape) + ":")
+        print(a_num)
+        assert False
+
+
+def generate_random(shape, datatype):
```
Comment (magnatelee) [inline tests/integration/test_searchsorted.py:31]: you call this function only with 1-tuples. can't you just pass volumes directly? if this is copied from some other test you wrote, consider making it a common utility. (at some point we will consolidate all these ad-hoc input generations into one module and make all tests use it.)

File: tests/integration/test_searchsorted.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+#
+
+import numpy as np
+import pytest
+
+import cunumeric as num
+
+
+def compare_assert(a_np, a_num):
+    if not num.allclose(a_np, a_num):
```
Comment (magnatelee) [inline tests/integration/test_searchsorted.py:23]: a nitpick: can't this be `array_equal`? I think `searchsorted` shouldn't ever do anything that could cause rounding errors

File: src/cunumeric/sort/searchsorted.cc [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                  const Rect<1>& rect_values,
+                  const bool left,
+                  const bool is_index_space,
+                  const size_t volume,
+                  const int64_t global_volume,
+                  const size_t num_values)
+  {
+    auto input   = input_array.read_accessor<VAL, 1>(rect_base);
+    auto input_v = input_values.read_accessor<VAL, 1>(rect_values);
+    assert(input.accessor.is_dense_arbitrary(rect_base));
```
Comment (magnatelee) [inline src/cunumeric/sort/searchsorted.cc:42]: please request an exact instance for the input to make this assertion hold all the time

**Comment by magnatelee** [comment]
@mfoerste4 doing a full search in every iteration seems pretty bad for memory bandwidth. do you know of any optimized algorithms that search for multiple values simultaneously? at minimum we could sort the search values to improve the access locality.

**Comment by mfoerste4** [comment]
> @mfoerste4 doing a full search in every iteration seems pretty bad for memory bandwidth. do you know of any optimized algorithms that search for multiple values simultaneously? at minimum we could sort the search values to improve the access locality.

I will bring this question up in the next tech-meeting. 

@ipdemes , could you do a second pass review?

**Comment by ipdemes** [comment]
@mfoerste4 there  are some corner cases I think we would like to test explicitly. (I think you do test them by generating inputs randomly, but, maybe, we would like to force these tests ):
1)some entries of `V` are not in the array `a`. 
2)some entries of `V` are the same
3) `v` is empty
4) `a` is empty

**Comment by ipdemes** [comment]
another comment for the test:
Please make sure you test this method for up to LEGATE_MAX_DIM array dimension. Please see an example here: 
https://github.com/nv-legate/cunumeric/blob/branch-22.07/tests/integration/test_diag_indices.py#L30_L37

File: src/cunumeric/mapper.cc [hunk 7]
```diff
@@ -161,6 +161,13 @@ std::vector<StoreMapping> CuNumericMapper::store_mappings(
       mappings.back().policy.exact = true;
       return std::move(mappings);
     }
+    case CUNUMERIC_SEARCHSORTED: {
+      std::vector<StoreMapping> mappings;
+      auto& inputs = task.inputs();
+      mappings.push_back(StoreMapping::default_mapping(inputs[0], options.front()));
```
Comment (ipdemes) [inline src/cunumeric/mapper.cc:167]: what does this do?

File: cunumeric/array.py [hunk 8]
```diff
@@ -3048,6 +3048,54 @@ def setflags(self, write=None, align=None, uic=None):
...
+
+        Returns
+        -------
+        indices : int or array_like[int]
+            Array of insertion points with the same shape as v, or an integer
+            if v is a scalar.
+
+        Availability
+        --------
+        Multiple GPUs, Multiple CPUs
```
Comment (ipdemes) [inline cunumeric/array.py:3084]: this is not required in the docstring of the cunuemeric/array,py file, only at the  cunumeric/module.py file

File: cunumeric/deferred.py [hunk 9]
```diff
@@ -1924,6 +1924,32 @@ def unique(self):
...
+        else:
+            self.fill(np.array(0, self.dtype))
+            task.add_reduction(self.base, ReductionOp.MAX)
+
+        task.add_input(rhs.base)
+        task.add_input(v.base)
+
+        # every partition needs the value information
+        task.add_broadcast(v.base)
+        task.add_broadcast(self.base)
```
Comment (ipdemes) [inline cunumeric/deferred.py:1946]: with this approach we map entire `v` and `self` to a single thread/task. Wile for the case when `v` is an array I don't necessarily see how can we implement this differently, we can do an optimization for the case when v is a scalar: in this case we can partition `self` and broadcast `v`

File: src/cunumeric/sort/searchsorted_template.inl [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+{
+  SearchSortedArgs args{context.inputs()[0],
+                        context.inputs()[1],
+                        context.reductions()[0],
+                        context.scalars()[0].value<bool>(),
+                        context.scalars()[1].value<int64_t>(),
+                        !context.is_single_task()};
+
+  assert(args.input_base.dim() == 1);
+  assert(args.input_base.code() == args.input_values.code());
```
Comment (ipdemes) [inline src/cunumeric/sort/searchsorted_template.inl:73]: I don't think that `input_base` and `input_values` are required to be the same type.
For example the example below works just fine in numpy:
```
arr = np.array([1, 2, 3, 4, 5], dtype=int)
arr2= np.array([1.0, 4.1], dtype = float)
res = np.searchsorted(arr, arr2)
```

File: cunumeric/array.py [hunk 11]
```diff
@@ -3048,6 +3048,54 @@ def setflags(self, write=None, align=None, uic=None):
...
+        """
+
```
Comment (ipdemes) [inline cunumeric/array.py:3086]: I think you might want to check that self is a 1D array somewhere here, if not, return the error.
I assume we would also want to check that V is a 1D array or scalar here as well

Comment (ipdemes) [reply cunumeric/array.py:3086]: Actually V can be a multidimensional array. This example works just fine:
```
arr = np.array([1, 2, 3, 4, 5], dtype=int)
arr2= np.array([[1.0, 4.1],[0.0, 3.2]], dtype = float)
res = np.searchsorted(arr, arr2)
```

File: src/cunumeric/mapper.cc (hunk 7 above)
Comment (mfoerste4) [reply src/cunumeric/mapper.cc:167]: The next line `exact=true` is supposed to ensure that we always have dense data to ensure
`assert(input.accessor.is_dense_arbitrary(rect_base));`

File: cunumeric/array.py (hunk 8 above)
Comment (mfoerste4) [reply cunumeric/array.py:3084]: I have seen it in some places in `array.py` as well (some of them are from myself though). Should I remove it? Just Availability or the whole doc?

File: cunumeric/deferred.py (hunk 9 above)
Comment (mfoerste4) [reply cunumeric/deferred.py:1946]: I don't really understand that. In case `v` is a scalar `self `should be a scalar as well, right? I just treat/convert scalars as ndarrays of size 1 for simplicity.

File: src/cunumeric/sort/searchsorted_template.inl (hunk 10 above)
Comment (mfoerste4) [reply src/cunumeric/sort/searchsorted_template.inl:73]: Oh, I did not think about that. I will check what kind of cross-dtypes are supported and will look for a solution to this.

File: cunumeric/array.py (hunk 11 above)
Comment (mfoerste4) [reply cunumeric/array.py:3086]: I added a check for self and sorter (if available).

File: src/cunumeric/sort/searchsorted_template.inl (hunk 10 above)
Comment (mfoerste4) [reply src/cunumeric/sort/searchsorted_template.inl:77]: I have added a `astype `conversion to a common `dtype `on python side. I have also added a followup task to improve this by making a double-type-dispatch and casting values within the kernels right after loading them.

**Comment by mfoerste4** [comment]
@ipdemes , I think I have addressed all of your suggestions. Please have another look. Thanks.

File: cunumeric/array.py [hunk 12]
```diff
@@ -3048,6 +3048,72 @@ def setflags(self, write=None, align=None, uic=None):
...
+        if self.ndim != 1:
+            raise ValueError("Dimension mismatch: self must be a 1D array")
+
```
Comment (ipdemes) [inline cunumeric/array.py:3089]: I think you are missing the case here when self.shape[0]==0:
```
arr = np.array([])
v = np.array([1, 2, 3, 4, 5])
print(arr.searchsorted(v))
```
I think the logic should be:
```
if a.shape[0] == 0:
        return zeros(v.shape, dtype=int64)
```

Comment (mfoerste4) [reply cunumeric/array.py:3089]: The case with empty self is tested, it works natively because the result-reduction is initialized with `0` or` self.size` (depending on `side`) before the task is started.

File: cunumeric/array.py [hunk 13]
```diff
@@ -3048,6 +3048,72 @@ def setflags(self, write=None, align=None, uic=None):
...
+            if v_ndarray.dtype is not ch_dtype:
+                v_ndarray = v_ndarray.astype(ch_dtype)
+            if a.dtype is not ch_dtype:
+                a = a.astype(ch_dtype)
+
+        if sorter is not None and a.shape[0] > 1:
+            if sorter.ndim != 1:
+                raise ValueError(
+                    "Dimension mismatch: sorter must be a 1D array"
+                )
```
[the remark this answers is not in the document]
Comment (mfoerste4) [reply cunumeric/array.py:3107]: I added 2 more checks for type and shape
